compatibility/deploy.go

```diff
@@ -0,0 +1,224 @@
+*/
+
+package compatibility
+
+import (
+	"fmt"
+
+	"github.com/compose-spec/compose-go/types"
+)
+
+func (c *AllowList) CheckDeploy(service *types.ServiceConfig) bool {
+	if !c.supported("deploy") && service.Deploy != nil {
+		service.Deploy = nil
+		c.error("deploy")
+		return false
+	}
+	return true
+}
+
+func (c *AllowList) CheckDeployMode(config *types.DeployConfig) {
+	if !c.supported("services.deploy.mode") && config.Mode != "" {
+		config.Mode = ""
+		c.error("services.deploy.mode")
+	}
+}
+func (c *AllowList) CheckDeployReplicas(config *types.DeployConfig) {
+	if !c.supported("services.deploy.replicas") && config.Replicas != nil {
+		config.Replicas = nil
+		c.error("services.deploy.replicas")
+	}
+}
+func (c *AllowList) CheckDeployLabels(config *types.DeployConfig) {
+	if !c.supported("services.deploy.labels") && len(config.Labels) != 0 {
+		config.Labels = nil
+		c.error("services.deploy.labels")
+	}
+}
+
+const (
+	UpdateConfigUpdate   = "update_config"
+	UpdateConfigRollback = "rolback_config"
+)
+
+func (c *AllowList) CheckDeployUpdateConfig(config *types.DeployConfig) bool {
+	if !c.supported("services.deploy.update_config") {
+		config.UpdateConfig = nil
+		c.error("services.deploy.update_config")
+		return false
+	}
+	return true
+}
+
+func (c *AllowList) CheckDeployRollbackConfig(config *types.DeployConfig) bool {
+	if !c.supported("services.deploy.rollback_config") {
+		config.RollbackConfig = nil
+		c.error("services.deploy.rollback_config")
+		return false
+	}
+	return true
+}
+
+func (c *AllowList) CheckUpdateConfigParallelism(s string, config *types.UpdateConfig) {
+	k := fmt.Sprintf("services.deploy.%s.parallelism", s)
+	if !c.supported(k) && config.Parallelism != nil {
+		config.Parallelism = nil
+		c.error(k)
+	}
+}
+func (c *AllowList) CheckUpdateConfigDelay(s string, config *types.UpdateConfig) {
+	k := fmt.Sprintf("services.deploy.%s.delay", s)
+	if !c.supported(k) && config.Delay != 0 {
+		config.Delay = 0
+		c.error(k)
+	}
+}
+func (c *AllowList) CheckUpdateConfigFailureAction(s string, config *types.UpdateConfig) {
+	k := fmt.Sprintf("services.deploy.%s.failure_action", s)
+	if !c.supported(k) && config.FailureAction != "" {
+		config.FailureAction = ""
+		c.error(k)
+	}
+}
+func (c *AllowList) CheckUpdateConfigMonitor(s string, config *types.UpdateConfig) {
+	k := fmt.Sprintf("services.deploy.%s.monitor", s)
+	if !c.supported(k) && config.Monitor != 0 {
+		config.Monitor = 0
+		c.error(k)
+	}
+}
+func (c *AllowList) CheckUpdateConfigMaxFailureRatio(s string, config *types.UpdateConfig) {
+	k := fmt.Sprintf("services.deploy.%s.max_failure_ratio", s)
+	if !c.supported(k) && config.MaxFailureRatio != 0 {
+		config.MaxFailureRatio = 0
+		c.error(k)
+	}
+}
+func (c *AllowList) CheckUpdateConfigOrder(s string, config *types.UpdateConfig) {
+	k := fmt.Sprintf("services.deploy.%s.order", s)
+	if !c.supported(k) && config.Order != "" {
+		config.Order = ""
+		c.error(k)
+	}
+}
+
+const (
+	ResourceLimits       = "limits"
```

Comment:
Where are these consts used?

Reply:
CheckDeployResourcesNanoCPUs(ResourceLimits, service.Deploy.Resources.Limits)
CheckDeployResourcesNanoCPUs(ResourceReservations, service.Deploy.Resources.Limits)

Golang do not support method override 😥